src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,743 @@
+          (hashUpdate (hashInitWith alg) (tableNameToHashInput tblName))
+          rows
+
+alg :: SHA3_256
+alg = SHA3_256
+
+tableNameToHashInput :: Text -> ByteString
+tableNameToHashInput tblName =
+  let
+    tbl = Text.encodeUtf8 (Text.toLower tblName)
+  in
+  BL.toStrict
+  $ BB.toLazyByteString
+  $ BB.charUtf8 'T'
+    <> BB.word64LE (fromIntegral @Int @Word64 (BS.length tbl))
+    <> BB.byteString tbl
+
+-- | Turn a (TableName, PactRow) into the arguments for a hash function.
+--
+--   The format is:
+--
+--   char 'K' (1 byte)
+--   <length rowkey> (8 bytes)
+--   <rowkey> (variable, maximum size TBD)
+--
+--   char 'I' (1 byte)
+--   <txId> (8 bytes)
+--
+--   char 'D' (1 byte)
+--   <rowdata> (variable, maximum size TBD)
+--
+--   And thus the size of this will always be 28 bytes + the total size of
+--   all variable sized inputs. None of the variable-sized inputs can
+--   be empty (other than maybe rowdata?).
+rowToHashInput :: PactRow -> ByteString
+rowToHashInput pr =
+  BL.toStrict
+  $ BB.toLazyByteString
+  $
+       BB.charUtf8 'K'
+    <> BB.word64LE (fromIntegral @Int @Word64 (BS.length pr.rowKey))
+    <> BB.byteString pr.rowKey
+
+    <> BB.charUtf8 'I'
+    <> BB.word64LE (fromIntegral @Int64 @Word64 pr.txId)
+
+    <> BB.charUtf8 'D'
+    <> BB.word64LE (fromIntegral @Int @Word64 (BS.length pr.rowData))
+    <> BB.byteString pr.rowData
+
+limitCut :: (Logger logger)
+  => logger
+  -> WebBlockHeaderDb
+  -> HashMap ChainId BlockHeader -- ^ latest cut headers
+  -> HashMap ChainId SQLiteEnv
+  -> BlockHeight
+  -> IO (HashMap ChainId BlockHeader)
+limitCut logger wbhdb latestCutHeaders pactConns blockHeight = do
+  fmap (HM.mapMaybe id) $ flip HM.traverseWithKey latestCutHeaders $ \cid latestCutHeader -> do
+    let logger' = addChainIdLabel cid logger
+    bdb <- getWebBlockHeaderDb wbhdb cid
+    seekAncestor bdb latestCutHeader (fromIntegral blockHeight) >>= \case
+      -- Block exists on that chain
+      Just h -> do
+        -- Sanity check, should absolutely never happen
+        when (_blockHeight h /= blockHeight) $ do
+          exitLog logger' "expected seekAncestor behaviour is broken"
+
+        -- Confirm that PactDB is not behind RocksDB (it can be ahead though)
+        let SQLiteEnv db _ = pactConns ^?! ix cid
+        latestPactHeight <- getLatestBlockHeight db
+        when (latestPactHeight < blockHeight) $ do
+          exitLog logger' "Pact State is behind RocksDB. This should never happen."
+
+        pure (Just h)
+
+      -- Block does not exist on that chain
+      Nothing -> do
+        logFunctionText logger' Debug $ "Block " <> sshow blockHeight <> " is not accessible on this chain."
+        pure Nothing
+
+resolveLatestCutHeaders :: (Logger logger)
+  => logger
+  -> ChainwebVersion
+  -> HashMap ChainId SQLiteEnv
+  -> RocksDb
+  -> IO (BlockHeight, HashMap ChainId BlockHeader)
+resolveLatestCutHeaders logger v pactConns rocksDb = do
```

Comment:
Here as well

Reply:
Yeah the difference between this and `readHighestCutHeaders` probably bears commenting.